src/edu/washington/escience/myria/expression/evaluate/PythonUDFEvaluator.java

```diff
@@ -0,0 +1,422 @@
+      DataOutputStream dOut = pyWorker.getDataOutputStream();
+      int numTuples = 0;
+      for (int j = 0; j < ltb.size(); j++) {
+        numTuples += ltb.get(j).numTuples();
+      }
+      pyWorker.sendNumTuples(numTuples);
+
+      for (int tbIdx = 0; tbIdx < ltb.size(); tbIdx++) {
+        TupleBatch tb = ltb.get(tbIdx);
+        for (int tup = 0; tup < tb.numTuples(); tup++) {
+          for (int col = 0; col < tupleSize; col++) {
+            writeToStream(tb, tup, columnIdxs[col], dOut);
+          }
+        }
+      }
+
+      // read result back
+      Object obj = readFromStream(null);
+      switch (outputType) {
+        case DOUBLE_TYPE:
+          result.putDouble(resultcol, (Double) obj);
+          break;
+        case BLOB_TYPE:
+          result.putBlob(resultcol, (ByteBuffer.wrap((byte[]) obj)));
+          break;
+        case FLOAT_TYPE:
+          result.putFloat(resultcol, (float) obj);
+          break;
+        case INT_TYPE:
+          result.putInt(resultcol, (int) obj);
+          break;
+        case LONG_TYPE:
+          result.putLong(resultcol, (long) obj);
+          break;
+
+        default:
+          LOGGER.info("type not supported as Python Output");
+          break;
+      }
+
+    } catch (Exception e) {
+
+      throw new DbException(e);
+    }
+  }
+
+  /**
+   * sendinput to be evaluated by python process.
+   * @param tb -input tuples.
+   * @param rowIdx -row to be evaluated.
+   * @param state - state column to be updated.
+   * @return Object - returned object.
+   * @throws IOException
+   * @throws DbException.
+   */
+  private Object evaluatePython(
+      final ReadableTable tb,
+      final int rowIdx,
+      final ReadableTable state,
+      final WritableColumn count)
+      throws DbException, IOException {
+    if (pyWorker == null) {
+      pyWorker = new PythonWorker();
+      initEvaluator();
+    }
+
+    try {
+      DataOutputStream dOut = pyWorker.getDataOutputStream();
+      pyWorker.sendNumTuples(1);
+      for (int i = 0; i < tupleSize; i++) {
+        if (isStateColumn[i]) {
+          writeToStream(state, rowIdx, columnIdxs[i], dOut);
+        } else {
+          writeToStream(tb, rowIdx, columnIdxs[i], dOut);
+        }
+      }
+
+      // read response back
+      Object result = readFromStream(count);
+      return result;
+
+    } catch (DbException e) {
+      LOGGER.info("Error writing to python stream" + e.getMessage());
+      throw new DbException(e);
+    }
+  }
+
+  /**
+   *
+   * @return Object output from python process.
+   * @throws DbException
+   */
+  private Object readFromStream(final WritableColumn count) throws DbException {
+    int type = 0;
+    Object obj = null;
+    DataInputStream dIn = pyWorker.getDataInputStream();
+    int c = 1;
+    try {
+      // read length of incoming message
+      type = dIn.readInt();
+      if (isFlatmap) {
+        c = dIn.readInt();
+        count.appendInt(c);
+      }
+
+      for (int i = 0; i < c; i++) {
```

Comment:
c should always be 1 for not flatmap, this works correctly. I have re-written to make logic cleaner.